src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var monitorSelectionService = (IVsMonitorSelection)await asyncServiceProvider.GetServiceAsync(typeof(SVsShellMonitorSelection));
+
+                if (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out var value)))
+                {
+                    if (value is IVsWindowFrame windowFrame)
+                    {
+                        TrackWindowFrame(windowFrame);
+                        _activeFrame = windowFrame;
+                    }
+                }
+
+                monitorSelectionService.AdviseSelectionEvents(this, out var _);
+            }));
+        }
+
+        public void Dispose()
+        {
+            _asyncTasks.Join();
+        }
+        
+        /// <summary>
+        /// Raised when the set of window frames being tracked changes, which means the results from <see cref="GetActiveDocument"/> or <see cref="GetVisibleDocuments"/> may change.
+        /// </summary>
+        public event EventHandler DocumentsChanged;
+
+        /// <summary>
+        /// Returns the <see cref="DocumentId"/> of the active document in a given <see cref="Workspace"/>.
+        /// </summary>
+        public DocumentId GetActiveDocument(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+            if (_activeFrame == null || visibleFramesSnapshot.IsEmpty)
+            {
+                return null;
+            }
+
+            foreach (var listener in visibleFramesSnapshot)
+            {
+                if (listener.Frame == _activeFrame)
+                {
+                    return listener.GetDocumentId(workspace);
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Get a read-only collection of the <see cref="DocumentId"/>s of all the visible documents in the given <see cref="Workspace"/>.
+        /// </summary>
+        public ImmutableArray<DocumentId> GetVisibleDocuments(Workspace workspace)
+        {
+            ThisCanBeCalledOnAnyThread();
+
+            var visibleFramesSnapshot = _visibleFrames;
+
+            var ids = ArrayBuilder<DocumentId>.GetInstance(visibleFramesSnapshot.Count);
+
+            foreach (var frame in visibleFramesSnapshot)
+            {
+                var documentId = frame.GetDocumentId(workspace);
+
+                if (documentId != null)
+                {
+                    ids.Add(documentId);
+                }
+            }
+
+            return ids.ToImmutableAndFree();
```

Comment:
passing in a .Count for the array builder is not a problme here right?  we still only get back an ImmutbaleArray with only the items we added?